dap_py/dap.py

```diff
@@ -0,0 +1,610 @@
+import os
+import json
+import base64
+from pathlib import Path
+import requests
+from getpass import getpass
+from .utils.scraper import get_api_url
+
+class BadStatusCodeError(RuntimeError):
+    def __init__(self, req):
+        self.status_code = req.status_code
+        try:
+            self.reason = req.json()['message']
+        except (KeyError, ValueError):
+            self.reason = req.reason
+
+    def __str__(self):
+        return ((
+            'Server Returned Bad Status Code\n'
+            f'Status Code: {self.status_code}\n'
+            f'Reason: {self.reason}')
+        )
+
+
+class dap:
+
+    def __init__(self, host_URL, cert_path=None, save_cert_dir=str(Path.home() / "dap/certs/"),
+                 download_dir = str(Path.home() / "dap/downloads"), quiet=False, spp=False):
+        '''initializes connection with DAP server and performs authentication
+
+        Args:
+            host_URL (str): The url of the host, e.g. "livewire.energy.gov"
+            cert (str, optional): path to authentication certificate file. Defaults to None.
+            save_cert_dir (str, optional): Path to directory where certificates are stored. Defaults to ~/dap/certs/
+            download_dir (str, optional): Path to directory where files will be downloaded. Defaults to ~/dap/downloads/
+            quiet (bool, optional): suppresses output print statemens. Useful for scripting Defaults to False.
+            spp (bool, optional): If this is a dap for the Solid Phase Processing data.
+        '''
+        self.host_URL = host_URL
+        if spp:
+            self._api_url = "https://13tl7mor8f.execute-api.us-west-2.amazonaws.com/prod"
+        else:
+            self._api_url = get_api_url(self.host_URL)
+
+        self._quiet = quiet
+        self._cert_path = cert_path
+        self._cert = None
+        self._auth = None
+
+        # set the certificate and download paths.
+        if "DAP_CERT_DIR" in os.environ:
+            self.save_cert_dir = os.environ["DAP_CERT_DIR"]
+        else:
+            self.save_cert_dir = save_cert_dir
+
+        if "DAP_DOWNLOAD_DIR" in os.environ:
+            self.download_dir = os.environ["DAP_DOWNLOAD_DIR"]
+        else:
+            self.download_dir = download_dir
+
+        self.__create_dirs()
+
+        self.__read_cert()
+
+        if self.__cert_is_valid():
+            self.__create_cert_auth()
+            self.__print('Certificate is setup')
+        else:
+            self.setup_guest_auth()
+            self.__print('Certificate not found or was invalid. Using guest credentials...')
+
+    def __print(self, *args, sep='', end='\n', file=None):
+        if not self._quiet:
+            for arg in args:
+                print(arg, sep=sep, end=end, file=file)
+
```

Comment:
Extraneous white space.  If you use `flake8` or some other linter, you can fix things like this really easily.  The convention is to have two blank lines between classes and functions but only one between methods of a class.

Reply:
I'd probably recommend using [`black`](https://github.com/psf/black) to handle code formatting. If you use VS Code you can set it up so that VS Code automatically uses `black` to format your file whenever you save. Once you do that you'll never have to think about formatting again :]